### ISSUE TITLE
Serialization and Reflection

### ISSUE THREAD

**Issue by DanVanAtta** [triplea-game/triplea#20]
Cool technologies, but they have some big problems:
- serialization means that we don't get compatibility when we do small things such as change names. This is brittle. A dependency on the name of a private variable is effectively making that variable public in scope.  This violation of encapsulation is probably largely what makes serialization so brittle. It's even worse than that since the dependency is saved on people's hard drives and remains for months, and cannot be changed in a backwards compatible manner.
- reflection, not only do we lose compatibility between game versions, but we also lose compile time checking that method calls and signatures match, this is brittle

Should we fix? Has this been discuss before? Current thoughts/ideas on how to fix?

**Comment by DanVanAtta** [comment]
It just occurred to me how to solve most of the problems. Specifically if we can save games without using serialization, then we avoid the long lived dependencies on specific serialized objects. When people save the game and quit the application then we lose the dependency on the serialized object. Today it lasts as long as a save game file lasts.

This also really helps the upgrade process. Old lobby is shut down, everyone joins new lobby and  happily loads save games.

**Comment by veqryn** [comment]
I'd be fine with this, but it is a lot of work and we probably have much higher priority items to address.
Obviously a big issue with this, is if we aren't going to use serialized objects, then what will we use in its place?

Also, this currently does not affect the upgrade process.  I made changes around 2 years ago (i think) so that all users join the same lobby (only 1 lobby at a time, for all users, all versions), but can only join games that are using the same engine version as the user has (this is why we have an 'EV' column, which stands for engine version).  Also, old savegames will load, even when we break serialization compatibility, because triplea will spin up a new process loading that game using an older jar to run it (this is why we have an "old" folder, with older jars in it).

**Comment by DanVanAtta** [comment]
We need to have a discussion about projects and priorities. I'll spin up another issue to discuss and hopefully a wiki page as well to start the list.

The EV feature did help quite a lot for the upgrade process. If we remove serialization and reflection I see two benefits from a user point of view that EV does not provide:  old user clients will be able to join newer lobbies. Secondly, new clients will be able to play any old save game (assuming the game parsing code is backwards compatible, but if not it can still have version specific  algorithms). Meaning we can drop the legacy engine support, since all we need to do is successfully parse the save game data in the new engine. Then we can either force people to join new lobbies with a new client, in which case we can also drop the EV check, or probably best is allow old clients to join new lobbies, in which case the EV feature is perfect.

There is some benefit there already. The most significant benefit though is if we can get down to compile time checking of all the code. For a moment I thought the code had only compile time checking and was getting pretty giddy, then found out about the serialization and reflection stuff pretty quickly and it turns out it's much more the extreme the other way.

Back to the topic at hand more directly, figuring out a good data format to save games in is a good open question. I'm not 100% sure, GSON maybe as a reasonable first easy choice. Still some more details to consider before we start honing in too specifically on exact persistence implementation.

**Comment by ron-murhammer** [comment]
So I can see why you grouped serialization and reflection together but these are pretty separate topics based on what part of the code they impact IMO. 

Serialization affects saves games so we would need to research what a better format would be a consider how much effort it would be to change it. This would seem like a major win though as we would most likely have much more control over save game compatibility and refactoring code.

Reflection I believe is mostly used in parsing the maps (there may be other areas as well but that's the main one I'm aware of). I personally don't like the way this code was written and would really prefer to use XML parsing frameworks and validation. This would mean most likely locking down more of the XML schema so options are more defined and less string parameters are provided. This would be a large change because you'd probably also end up affecting all map XMLs so would need to write a utility to convert them.

**Comment by veqryn** [comment]
serialization not just affects savegames, but also all communication over the network (ie: dice, "change" objects, politics, chats, commands to run a sound, etc)

honestly i am not sure getting rid of serialization would buy us that much, as that would just mean we are forced to maintain a very specific API for what a gamesave and other stuff is stored as (ie: json or xml, etc), and this would be constantly changing as we add / change / rename features/properties

reflection is used not just in parsing a map into a game data object, but also is used by our network communication

**Comment by DanVanAtta** [comment]
I mentioned reflection largely since that is used as part of the RMI calls which do the networking stuff. RMI requires serializable objects.

@veqryn , serialization is as specific of an API as it gets. Even worse than just being a specific API, the bloody API is saved to every users hard disk and then persisted there for an arbitrary amount of time. With serialization if we change the name of a field, as you mentioned it would break save games. This API would at least let us change the name of private fields, but as you pointed out there would still be elements of coupling. Plus there are all of the game engine benefits to not having serialized save games, helps upgrade etc, and lastly the API would be specific maybe, but at least it would be explicit. Specifically serialization is a bit error prone, there is a lot of auto-magic stuff that happens, lots of hard to understand error cases, so removing it where we can is a win both in terms of simplicity and reducing coupling.

**Comment by ron-murhammer** [comment]
While I agree having a sort of API to replace serialization would be preferable, this would be a ton of work which probably makes it low priority. 

Personally, I think dealing with the map parsing would be less work and have a bigger impact.

**Comment by DanVanAtta** [comment]
Just because an element of work is large does not diminish priority. Benefit to cost ratio is impacted, but the benefit is big.
We have some crazy APIs that we are stuck with, methods with many boolean arguments, half a dozen or more arguments, sometimes half of them are not used. And we can't fix it easily with 100% reliance on serialization.

I think the perils of serialization probably have been thoroughly discussed. Getting rid of it everywhere is not practical. Doing so as a first step for map parsing and saving I think is a reasonable chunk that gives us some good benefit. We can then look at improving the network stuff which would then let us change code with less risk (and also avoid version incompatibility problems, allow old game engines to run in new game lobbies)

**Comment by DanVanAtta** [comment]
Thought of some more pretty good benefits of doing our own save API:
- we could have AI persist data in saves
- we could auto-save during battles

v1.0 may not have those features, but my thinking is it'll be more straight forward to have a specific API rather than rely on serialization (which is an API of its own as well)

**Comment by veqryn** [comment]
We already save during battles...
You (and everyone else) may want to spend a bit more time digging around the code before we undertake something like this.  Why don't we revisit this discussion in several months?

**Comment by DanVanAtta** [comment]
We save between battles, but not during? Like we don't save after every dice roll? I saw a comment in the code that stated that we couldn't autosave at that time.

The couple months I think is arbitrary, but figuring out which are the right next infrastructure projects to tackle I think is a good thing to do. Then when we have enough labor/skill, my opinion is that this likely will be that next big infrastructure project after improving testing and all the dev-ops type of stuff we've been doing.

**Comment by veqryn** [comment]
We do save after every single atomic phase of the battle, which is broken down as much as possible so that we basically save after every die or group of dice is rolled.
